crates/interfaces/src/p2p/full_block_range.rs

```diff
@@ -0,0 +1,237 @@
+    fmt::Debug,
+    future::Future,
+    pin::Pin,
+    task::{ready, Context, Poll},
+};
+use tracing::debug;
+
+use super::headers::client::HeadersRequest;
+
+/// A Client that can fetch full blocks from the network.
+#[derive(Debug, Clone)]
+pub struct FullBlockRangeClient<Client> {
+    client: Client,
+}
+
+impl<Client> FullBlockRangeClient<Client> {
+    /// Creates a new instance of `FullBlockRangeClient`.
+    pub fn new(client: Client) -> Self {
+        Self { client }
+    }
+}
+
+impl<Client> FullBlockRangeClient<Client>
+where
+    Client: BodiesClient + HeadersClient + Clone,
+{
+    /// Returns a future that fetches [SealedBlock]s for the given hash and count.
+    ///
+    /// Note: this future is cancel safe
+    ///
+    /// Caution: This does no validation of body (transactions) responses but guarantees that
+    /// [SealedHeader]s match the requested hash.
+    pub fn get_full_block_range(
+        &self,
+        hash: H256,
+        count: u64,
+    ) -> FetchFullBlockRangeFuture<Client> {
+        let client = self.client.clone();
+
+        FetchFullBlockRangeFuture {
+            hash,
+            count,
+            request: FullBlockRangeRequest {
+                headers: Some(client.get_headers(HeadersRequest {
+                    start: hash.into(),
+                    limit: count,
+                    direction: HeadersDirection::Falling,
+                })),
+                bodies: None,
+            },
+            client,
+            headers: None,
+            bodies: None,
+        }
+    }
+}
+
+/// A future that downloads a range of full blocks from the network.
+#[must_use = "futures do nothing unless polled"]
+pub struct FetchFullBlockRangeFuture<Client>
+where
+    Client: BodiesClient + HeadersClient,
+{
+    client: Client,
+    hash: H256,
+    count: u64,
+    request: FullBlockRangeRequest<Client>,
+    headers: Option<Vec<SealedHeader>>,
+    bodies: Option<Vec<BlockBody>>,
+}
+
+impl<Client> FetchFullBlockRangeFuture<Client>
+where
+    Client: BodiesClient + HeadersClient,
+{
+    /// Returns the block hashes for the given range, if they are available.
+    pub fn range_block_hashes(&self) -> Option<Vec<H256>> {
+        self.headers.as_ref().map(|h| h.iter().map(|h| h.hash()).collect::<Vec<_>>())
+    }
+
+    /// Returns the [SealedBlock]s if the request is complete.
+    fn take_blocks(&mut self) -> Option<Vec<SealedBlock>> {
+        if self.headers.is_none() || self.bodies.is_none() {
+            return None
+        }
+
+        let headers = self.headers.take().unwrap();
+        let bodies = self.bodies.take().unwrap();
+        Some(
+            headers
+                .iter()
+                .zip(bodies.iter())
+                .map(|(h, b)| SealedBlock::new(h.clone(), b.clone()))
+                .collect::<Vec<_>>(),
+        )
+    }
+}
+
+impl<Client> Future for FetchFullBlockRangeFuture<Client>
+where
+    Client: BodiesClient + HeadersClient + Unpin + 'static,
+{
```

Comment:
The current implementation sends a single request for headers, and a single request for bodies, so there is only one thing to yield (the headers response + bodies response as full blocks).

A stream is definitely a better fit, especially if we want to send multiple requests for bodies. Because we don't do this yet, a `Future` works, but I'll add a `Stream` for the current implementation that just yields the full blocks in order.

Reply:
Just implemented the stream as `FullBlockRangeStream`, naming for all of these could use some work, but this basically first polls the future, and when the future is `Ready`, it buffers the result. Successive polls on the stream pop the buffered blocks until there are none left.